param/parameterized.py

```diff
@@ -1167,31 +1167,32 @@ def _watch(self_,action,watcher,parameter_name,parameter_attribute=None):
                 watchers[parameter_attribute] = []
             getattr(watchers[parameter_attribute],action)(watcher)
 
-    def watch(self_,fn,parameter_name,parameter_attribute=None):
-        watcher = Watcher(fn=fn, mode='args')
+    def watch(self_,fn,parameter_name,parameter_attribute=None, onlychanged=True):
+        watcher = Watcher(fn=fn, mode='args', onlychanged=onlychanged)
         self_._watch('append',watcher,parameter_name,parameter_attribute)
 
     def unwatch(self_,fn,parameter_name,parameter_attribute=None):
         """
         Unwatch watchers set either with watch or watch_values.
         """
         unwatched = False
-        try:
-            watcher = Watcher(fn=fn, mode='args')
-            self_._watch('remove',watcher,parameter_name,parameter_attribute)
-            unwatched = True
-        except: pass
-        try:
-            watcher = Watcher(fn=fn, mode='kwargs')
-            self_._watch('remove',watcher,parameter_name,parameter_attribute)
-            unwatched = True
-        except: pass
+        for onlychanged in [True, False]:
```

Comment:
Are the watchers hashed based on their onlychanged value or why is this necessary?

Reply:
Okay I see it, I'm happy to merge now but maybe we could consider changing the ``watchers`` to be a nested `dict` with keys of the form ``watchers[parameter_name][(parameter_attribute, fn)]``. That way you don't need to replicate the whole watcher state on unwatch.

Reply:
This is so named tuples match when removing from a list. If the value is different for `onlychanged` then it won't match. The alternative is to specify `onlychanged` as an argument which will only remove that type.